### DISCUSSION TITLE
[Enhancement] [button] el-button's gap should place at the end instead of front

### DISCUSSION THREAD

**Discussion by KyleBing** [element-plus/element-plus#16958] [Feature Request]
Feature type: **`Enhancement`**

## Details

- Enhanced component name: `button`

## Description

1. el-button's margin should place in the end instead of front, cause it will align uneven when it wrap to next line.
2. the end margin should use :last-child to process.

### Additional comments

![2024-05-23 15 01 47](https://github.com/element-plus/element-plus/assets/12215982/5f234526-58c0-4271-bbbc-9b67c4dbf65f)

![2024-05-23 15 02 24](https://github.com/element-plus/element-plus/assets/12215982/dc296d03-6df7-4e20-ac50-cc443929bac8)

**Comment by xxholly32** [comment]
Can you provide a replay demo

**Reply by KyleBing** [reply]
[demo](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6IjxzY3JpcHQgc2V0dXAgbGFuZz1cInRzXCI+XG5pbXBvcnQgeyByZWYsIHZlcnNpb24gYXMgdnVlVmVyc2lvbiB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IHZlcnNpb24gYXMgZXBWZXJzaW9uIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuaW1wb3J0IHsgRWxlbWVudFBsdXMgfSBmcm9tICdAZWxlbWVudC1wbHVzL2ljb25zLXZ1ZSdcblxuY29uc3QgdGFibGVEYXRhID0gcmVmKFtcbntpZDogMSwgbmFtZTogXCJwcmVzZW50XCIsIGRlbW86ICcnfSxcbntpZDogMiwgbmFtZTogXCJ3aXNoIHRvIGhhcHBlbmRcIiwgZGVtbzogJyd9LFxuXSlcbjwvc2NyaXB0PlxuXG5cbjx0ZW1wbGF0ZT5cbjxlbC10YWJsZSA6ZGF0YT1cInRhYmxlRGF0YVwiPlxuICAgIDxlbC10YWJsZS1jb2x1bW4gbGFiZWw9XCJkZW1vXCIgcHJvcD1cIm5hbWVcIiAvPlxuICAgIDxlbC10YWJsZS1jb2x1bW4gbGFiZWw9XCJidG5zXCIgcHJvcD1cImRlbW9cIj5cbiAgICAgIDx0ZW1wbGF0ZSAjZGVmYXVsdD1cIntyb3d9XCI+XG4gICAgICA8ZGl2IHYtaWY9XCJyb3cuaWQgPT09IDFcIj5cbiAgICAgICAgPGVsLWJ1dHRvbiBzaXplPVwic21hbGxcIiB0eXBlPVwicHJpbWFyeVwiIHYtZm9yPVwiaXRlbSBpbiA2XCIgOmtleT1cIml0ZW1cIj5CdXR0b24te3tpdGVtfX08L2VsLWJ1dHRvbj5cbiAgICAgICAgPHNwYW4gY2xhc3M9XCJzdG9wLXNpZ25cIj5FT0w8L3NwYW4+XG4gICAgICA8L2Rpdj5cbiAgICAgIDxkaXYgdi1lbHNlPlxuICAgICAgICA8ZWwtYnV0dG9uIGNsYXNzPVwiZml4ZWQtYnV0dG9uXCIgc2l6ZT1cInNtYWxsXCIgdHlwZT1cInByaW1hcnlcIiB2LWZvcj1cIml0ZW0gaW4gNlwiIDprZXk9XCJpdGVtXCI+QnV0dG9uLXt7aXRlbX19PC9lbC1idXR0b24+XG4gICAgICAgIDxzcGFuIGNsYXNzPVwic3RvcC1zaWduXCI+RU9MPC9zcGFuPlxuICAgICAgPC9kaXY+XG4gICAgICA8L3RlbXBsYXRlPlxuICAgIDwvZWwtdGFibGUtY29sdW1uPlxuXG48L2VsLXRhYmxlPlxuXG48L3RlbXBsYXRlPlxuXG5cbjxzdHlsZSAgc2NvcGVkPlxuLmZpeGVkLWJ1dHRvbntcbiAgbWFyZ2luLWxlZnQ6IDAgIWltcG9ydGFudDtcbiAgbWFyZ2luLXJpZ2h0OiAxMHB4O1xufVxuXG4uZml4ZWQtYnV0dG9uOmxhc3QtY2hpbGR7XG4gIG1hcmdpbi1yaWdodDogMCAhaW1wb3J0YW50O1xufVxuXG4uc3RvcC1zaWdue1xuICBjb2xvcjogZ3JlZW5cbn1cbjwvc3R5bGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJfbyI6e319)


by the way, it seems the demo website doesn't support `:last-child`, but `:nth-child(2n)` workes just fine.

```css
.fixed-button:last-child{
  margin-right: 0 !important;
}
```

**Reply by xxholly32** [reply]
![image](https://github.com/element-plus/element-plus/assets/6063358/0ae8d13a-7935-4dda-ac22-120c16af9475)

This is a history problem-design, which is how element-ui was originally designed

I don't like this mandatory setting, I remove it in my projects and implement the spacing myself; Use the flex-gap approach to this

[like this](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6IjxzY3JpcHQgc2V0dXAgbGFuZz1cInRzXCI+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG48L3NjcmlwdD5cblxuXG48dGVtcGxhdGU+XG48ZGl2IGNsYXNzPVwiZmxleCBmbGV4LXJvdyBmbGV4LXdyYXAgZ2FwLTQgdy0xMDBcIj5cbiAgPGVsLWJ1dHRvbiBzaXplPVwic21hbGxcIiB0eXBlPVwicHJpbWFyeVwiIHYtZm9yPVwiaXRlbSBpbiA2XCIgOmtleT1cIml0ZW1cIj5CdXR0b24te3tpdGVtfX08L2VsLWJ1dHRvbj5cbiAgPHNwYW4gY2xhc3M9XCJzdG9wLXNpZ25cIj5FT0w8L3NwYW4+XG48L2Rpdj5cblxuPC90ZW1wbGF0ZT5cblxuXG48c3R5bGUgIHNjb3BlZD5cbi5lbC1idXR0b24rLmVsLWJ1dHRvbntcbiAgbWFyZ2luOiAwO1xufVxuXG4uc3RvcC1zaWdue1xuICBjb2xvcjogZ3JlZW5cbn1cbjwvc3R5bGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJfbyI6e319)

**Comment by kooriookami** [comment]
I think the button shouldn't have the margin itself. If users need the gap, they can use flex gap, or add margin manually.
